File: Maximum Total Damage With Spell Casting/Maximum Total Damage With Spell Casting.java
// Solution for Maximum Total Damage With Spell Casting in JAVA

import java.util.*;

class Solution {
    // Recursive helper function to compute maximum total damage
    private long computeMaxDamage(List<Long> uniqueDamages, int i, long[] memo, Map<Integer, Long> damageCountMap) {
        // Base case: no more spells
        if (i >= uniqueDamages.size()) return 0;

        // If already computed, return memoized value
        if (memo[i] != -1) return memo[i];

        // Option 1: Skip current spell
        long skipCurrent = computeMaxDamage(uniqueDamages, i + 1, memo, damageCountMap);

        // Option 2: Take current spell damage and skip incompatible ones
        long currentDamage = uniqueDamages.get(i);
        long nextValidDamage = currentDamage + 3;

        // Binary search for next valid index (similar to upper_bound in C++)
        int nextIndex = Collections.binarySearch(uniqueDamages, nextValidDamage + 0L);
        if (nextIndex < 0) nextIndex = -(nextIndex + 1);

        long takeCurrent = currentDamage * damageCountMap.get((int) currentDamage)
                + computeMaxDamage(uniqueDamages, nextIndex, memo, damageCountMap);

        // Store and return the maximum of both options
        return memo[i] = Math.max(skipCurrent, takeCurrent);
    }

    public long maximumTotalDamage(int[] power) {
        Map<Integer, Long> damageCountMap = new HashMap<>();

        // Count frequency of each unique damage value
        for (int damage : power) {
            damageCountMap.put(damage, damageCountMap.getOrDefault(damage, 0L) + 1);
        }

        // Extract unique damage values and sort them
        List<Long> uniqueDamages = new ArrayList<>();
        for (int dmg : damageCountMap.keySet()) {
            uniqueDamages.add((long) dmg);
        }
        Collections.sort(uniqueDamages);

        int n = uniqueDamages.size();
        long[] memo = new long[n];
        Arrays.fill(memo, -1);

        // Compute and return the maximum possible total damage
        return computeMaxDamage(uniqueDamages, 0, memo, damageCountMap);
    }
}

File: Maximum Total Damage With Spell Casting/Maximum Total Damage With Spell Casting.py
# Solution for Maximum Total Damage With Spell Casting in PY

class Solution:
    def computeMaxDamage(self, unique_damages: List[int], i: int, memo: List[int], damage_count_map: dict) -> int:
        # Base case
        if i >= len(unique_damages):
            return 0

        # Return memoized result
        if memo[i] != -1:
            return memo[i]

        # Option 1: Skip current damage
        skip_current = self.computeMaxDamage(unique_damages, i + 1, memo, damage_count_map)

        # Option 2: Take current damage and skip incompatible ones
        current_damage = unique_damages[i]
        next_index = bisect.bisect_right(unique_damages, current_damage + 2)

        take_current = current_damage * damage_count_map[current_damage] + \
                       self.computeMaxDamage(unique_damages, next_index, memo, damage_count_map)

        # Store result
        memo[i] = max(skip_current, take_current)
        return memo[i]

    def maximumTotalDamage(self, power: List[int]) -> int:
        # Count occurrences of each damage value
        damage_count_map = {}
        for dmg in power:
            damage_count_map[dmg] = damage_count_map.get(dmg, 0) + 1

        # Extract and sort unique damages
        unique_damages = sorted(damage_count_map.keys())

        memo = [-1] * len(unique_damages)

        # Compute result
        return self.computeMaxDamage(unique_damages, 0, memo, damage_count_map)
